<a href="https://colab.research.google.com/github/Ranjitha-Raju/projects/blob/main/heart_knn_linkedin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


**Implementation of K-Nearest Neighbour on Heart disease dataset. I have used "heart.csv" data set for this.**

In [ ]:
#1. Importing the libraries and reading the dataset.
import numpy as np
import pandas as pd
df=pd.read_csv("/content/heart.csv")
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [ ]:
2.#checking for missing valuues
df.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [ ]:
#3.checking whether the data is balanced or imbalanced.
target_count=df.groupby('target') ['target'].count()
target_count

target
0    499
1    526
Name: target, dtype: int64

In [ ]:
#4.collecting input data.
x=df.iloc[:,:-1].values
x

array([[52.,  1.,  0., ...,  2.,  2.,  3.],
       [53.,  1.,  0., ...,  0.,  0.,  3.],
       [70.,  1.,  0., ...,  0.,  0.,  3.],
       ...,
       [47.,  1.,  0., ...,  1.,  1.,  2.],
       [50.,  0.,  0., ...,  2.,  0.,  2.],
       [54.,  1.,  0., ...,  1.,  1.,  3.]])

In [ ]:
#5.collecting the output data
y=df.iloc[:,-1].values
y

array([0, 0, 0, ..., 0, 1, 0])

In [ ]:
#6.I have split the dataset into training dataset and testing dataset.
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.30,random_state=42)
print(x_train)

[[59.  1.  1. ...  2.  0.  2.]
 [58.  1.  0. ...  1.  3.  3.]
 [44.  0.  2. ...  1.  1.  2.]
 ...
 [51.  1.  0. ...  2.  0.  3.]
 [43.  1.  0. ...  2.  0.  3.]
 [52.  1.  0. ...  2.  1.  2.]]


In [ ]:
print(y_train)

[1 0 1 0 1 1 0 0 1 1 0 1 0 1 1 1 0 0 1 0 0 1 1 1 1 1 0 0 1 1 1 1 1 0 1 1 1
 0 0 0 1 0 1 1 1 0 0 1 0 0 1 1 1 0 1 1 1 1 1 1 0 0 0 1 0 1 1 1 1 0 0 1 1 1
 0 1 1 1 0 0 1 0 0 0 1 1 1 1 1 0 0 1 0 1 1 0 0 1 0 0 0 0 1 0 0 1 1 0 1 0 0
 1 0 0 0 1 1 1 1 0 1 1 0 1 0 1 0 0 0 0 1 1 1 0 0 0 0 1 1 1 0 1 0 1 1 1 0 1
 1 0 1 1 1 0 0 0 1 0 1 0 1 1 0 1 1 1 1 1 1 0 0 0 1 0 1 0 0 1 0 1 1 1 0 0 0
 0 0 0 1 0 1 1 1 0 0 1 0 1 0 0 0 0 0 0 1 0 0 1 0 1 1 0 1 1 0 1 0 1 0 1 1 1
 1 1 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 1 1 0 1 1 1 1 1 1 0 1 0 0 0 0 1 0 1 0 1
 0 1 1 1 0 1 1 0 1 0 1 1 1 0 1 1 1 0 1 1 1 0 1 0 1 0 1 1 0 0 1 1 0 1 1 1 0
 1 0 1 1 1 0 1 0 0 1 1 0 1 0 1 0 0 1 0 1 0 0 1 0 0 0 0 1 0 0 0 0 1 1 1 1 1
 1 0 0 1 0 1 0 1 0 0 0 1 0 0 1 1 0 0 0 0 0 1 1 1 0 0 1 0 0 0 0 0 1 0 0 0 1
 0 1 0 0 0 1 1 0 1 1 1 1 0 0 1 1 1 0 0 1 1 1 0 1 1 0 0 1 0 1 0 0 0 0 0 1 1
 1 0 0 0 0 1 1 0 0 1 1 1 0 1 1 0 0 1 0 1 1 0 1 1 0 1 1 1 0 1 1 0 0 1 1 1 0
 1 1 0 0 0 0 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0 0 1 0 0 0 1 0 0 0 1 0 1 0 0 0 1
 0 1 0 0 1 1 1 0 0 0 0 1 

In [ ]:
#7.Dataset scaling(Normalization)
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
scaler.fit(x_train)
x_train=scaler.transform(x_train)
x_test=scaler.transform(x_test)
x_train

array([[ 0.51227152,  0.64358464,  0.02931827, ...,  0.97708096,
        -0.72837545, -0.53812812],
       [ 0.40223597,  0.64358464, -0.92619093, ..., -0.65214474,
         2.10991368,  1.10373511],
       [-1.1382618 , -1.55379719,  0.98482748, ..., -0.65214474,
         0.21772092, -0.53812812],
       ...,
       [-0.36801292,  0.64358464, -0.92619093, ...,  0.97708096,
        -0.72837545,  1.10373511],
       [-1.24829735,  0.64358464, -0.92619093, ...,  0.97708096,
        -0.72837545,  1.10373511],
       [-0.25797736,  0.64358464, -0.92619093, ...,  0.97708096,
         0.21772092, -0.53812812]])

In [ ]:
print(x_test)

[[ 0.84237819 -1.55379719 -0.92619093 ...  0.97708096 -0.72837545
  -0.53812812]
 [-0.14794181 -1.55379719  0.98482748 ...  0.97708096 -0.72837545
  -3.82185458]
 [ 0.0721293   0.64358464 -0.92619093 ... -0.65214474  0.21772092
   1.10373511]
 ...
 [ 1.72266262  0.64358464  0.98482748 ... -0.65214474  0.21772092
   1.10373511]
 [ 1.39255596  0.64358464 -0.92619093 ... -0.65214474  1.1638173
  -0.53812812]
 [ 1.0624493   0.64358464  0.98482748 ...  0.97708096 -0.72837545
  -0.53812812]]


In [ ]:
#8.Implementing K-Neighbours classifier and prediction on x_test dataset.
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=7)
knn.fit(x_train,y_train)
y_pred=knn.predict(x_test)
y_pred

array([1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0,

In [ ]:
#9.Evaluation of our K-NN algorithm.
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
matrix=confusion_matrix(y_test,y_pred)
print("Confusion Matrix:\n",matrix)
score=accuracy_score(y_test,y_pred)
print("Accuracy Score:",score)

Confusion Matrix:
 [[127  32]
 [ 17 132]]
Accuracy Score: 0.8409090909090909


In [ ]:
report=classification_report(y_test,y_pred)
print("Classification Report:\n",report)

Classification Report:
               precision    recall  f1-score   support

           0       0.88      0.80      0.84       159
           1       0.80      0.89      0.84       149

    accuracy                           0.84       308
   macro avg       0.84      0.84      0.84       308
weighted avg       0.84      0.84      0.84       308



The K-NN algorithm was able to classify the X_test set records with 84% accuracy.